core/src/main/java/com/sequenceiq/cloudbreak/service/cluster/ambari/AmbariClusterModificationService.java

```diff
@@ -89,60 +91,80 @@ public void upscaleCluster(Stack stack, HostGroup hostGroup, Collection<HostMeta
+            }
+
         }
     }
 
     @Override
     public void stopCluster(Stack stack) throws CloudbreakException {
         AmbariClient ambariClient = clientFactory.getAmbariClient(stack, stack.getCluster());
         try {
-            boolean stopped = true;
-            Collection<Map<String, String>> values = ambariClient.getHostComponentsStates().values();
-            for (Map<String, String> value : values) {
-                for (String state : value.values()) {
-                    if (!"INSTALLED".equals(state)) {
-                        stopped = false;
-                    }
-                }
-            }
-            if (!stopped) {
-                LOGGER.info("Stop all Hadoop services");
-                eventService
-                        .fireCloudbreakEvent(stack.getId(), UPDATE_IN_PROGRESS.name(),
-                                cloudbreakMessagesService.getMessage(AMBARI_CLUSTER_SERVICES_STOPPING.code()));
-                int requestId = ambariClient.stopAllServices();
-                if (requestId != -1) {
-                    LOGGER.info("Waiting for Hadoop services to stop on stack");
-                    PollingResult servicesStopResult = ambariOperationService.waitForOperations(stack, ambariClient, singletonMap("stop services", requestId),
-                            STOP_AMBARI_PROGRESS_STATE).getLeft();
-                    if (isExited(servicesStopResult)) {
-                        throw new CancellationException("Cluster was terminated while waiting for Hadoop services to start");
-                    } else if (isTimeout(servicesStopResult)) {
-                        throw new CloudbreakException("Timeout while stopping Ambari services.");
-                    }
-                } else {
-                    LOGGER.warn("Failed to stop Hadoop services.");
-                    throw new CloudbreakException("Failed to stop Hadoop services.");
+            if (!isClusterStopped(ambariClient)) {
+                try {
+                    stopHadoopServices(stack, ambariClient);
+                } catch (IOException | URISyntaxException e) {
+                    throw new CloudbreakException("Failed to stop Hadoop services.", e);
                 }
-                eventService
-                        .fireCloudbreakEvent(stack.getId(), UPDATE_IN_PROGRESS.name(),
-                                cloudbreakMessagesService.getMessage(AMBARI_CLUSTER_SERVICES_STOPPED.code()));
             }
         } catch (AmbariConnectionException ignored) {
             LOGGER.debug("Ambari not running on the gateway machine, no need to stop it.");
         }
     }
 
+    private boolean isClusterStopped(AmbariClient ambariClient) {
+        Collection<Map<String, String>> values = ambariClient.getHostComponentsStates().values();
+        for (Map<String, String> value : values) {
+            for (String state : value.values()) {
+                if (!"INSTALLED".equals(state)) {
```

Comment:
maybe it is worth to rewrite to stream filter and check count, isn't it? i hope the size of the components is not that big which can cause performance issues

Reply:
Well, actually I just extracted this method, and I didn't bother to rewrite it. It is readable, easy to understand and performant.

Reply:
Ok, we can leave it as-is, if you prefer that